### ISSUE TITLE
Add SingI (& other) pattern synonyms

### ISSUE THREAD

**Issue by Icelandjack** [goldfirere/singletons#273]
This would be useful

```haskell
pattern SingI :: () => SingI a => Π(a::k)
pattern SingI <- (singInstance -> SingInstance)
  where SingI = sing
```

----

Similarly I think the same should be added to [`Type.Reflection`](https://downloads.haskell.org/~ghc/master/libraries/html/base/Type-Reflection.html), where is best to propose that again?

```haskell
asTypeable :: TypeRep a -> Dict (Typeable a)
asTypeable rep =
  withTypeable rep
    Dict

pattern TypeRep :: () => Typeable a => TypeRep a
pattern TypeRep <- (asTypeable -> Dict)
  where TypeRep = typeRep
```

**Comment by RyanGlScott** [comment]
I must be missing something obvious here, since I don't see what utility this provides, given that every `Sing` data family instance comes patterns (in the form of constructors) that accomplish this very thing. Can you give an example that demonstrates where this would be a win?

**Comment by Icelandjack** [comment]
Anywhere where `withSingI` is useful (i.e. where you want to constraintify an explicit witness), using the continuation passing `with*` functions work fine but explicit pattern matching can lead to nicer code. 

These can be used as building blocks for other pattern synonyms.

```haskell
-- repl :: forall a n. SingI n => a -> Vec n a
-- repl a = case sing :: Π(n) of
--   SO    -> VO
--   SS n' -> a :> withSingI n' (repl a)

repl :: forall a n. SingI n => a -> Vec n a
repl a = case sing :: Π(n) of
  SO       -> VO
  SS SingI -> a :> repl a
```

Similarly any of these may be useful

### `AsSing`

```haskell
asSing :: SingKind k => Demote k -> SomeSing k
asSing demote =
  withSomeSing demote
    SomeSing

pattern AsSing :: SingKind k => forall (a :: k). Π(a::k) -> Demote k
pattern AsSing sing <- (asSing -> SomeSing sing)
  where AsSing sing = fromSing sing
```

uses

```haskell
-- toSing :: N -> SomeSing N
-- toSing O     = SomeSing SO
-- toSing (S n) = withSomeSing n (\n' -> SomeSing $ SS n')
toSing_ :: N -> SomeSing N
toSing_ = \case
  O -> SomeSing SO
  S (AsSing sn) ->
    SomeSing (SS sn)
```

### `Demoted`

```haskell
pattern Demoted :: SingKind k => Demote k -> Π(a::k)
pattern Demoted demote <- (fromSing -> demote)

neg :: Π (a::Bool) -> Bool
neg (Demoted bool) = not bool
```

**Comment by RyanGlScott** [comment]
Ah, now I see what you're getting at. You're not looking to supplant the `Sing` constructors at all, but rather, you want to reduce some boilerplate that `withSomeSing`, `singInstance`, and `fromSing` entail. That seems quite reasonable, then.

Care to submit a pull request? :)

**Comment by goldfirere** [comment]
I like `SingI` -- nice use of pattern synonyms.

**Comment by RyanGlScott** [comment]
The `SingI` pattern synonym was added in #276.

@Icelandjack, is there still more to be done here, or can this issue be closed?

**Comment by Icelandjack** [comment]
It depends, do the other patterns merit inclusion?

**Comment by RyanGlScott** [comment]
`AsSing` definitely has the same flavor of continuation-passing-style-as-patterns that `SingI` has, so that certainly seems like something folks might want.

I don't have an opinion on `Demoted`.

**Comment by Icelandjack** [comment]
I agree, will make a PR for `AsSing`.

**Comment by Icelandjack** [comment]
Also I found the wiki confusing: If I want to propose `TypeRep` (should probably be named `Typeable`) to `Type.Reflection` must I go through a mailing list?

**Comment by RyanGlScott** [comment]
Yes, additions to core libraries such as `base` are generally proposed on the [Haskell libraries mailing list](https://mail.haskell.org/mailman/listinfo/libraries).

**Comment by RyanGlScott** [comment]
I believe this is resolves as of #277.
